### DISCUSSION TITLE
Morphometrics extraction intepretation and segmentation issue

### DISCUSSION THREAD

**Discussion by Paasu** [neuropoly/axondeepseg#329] [Help needed] [answered]
I am having trouble understanding the results of the morphometrics. What does the myelin diameter mean vs. the myelin thickness in this regard? Additionally, the myelin thickness seems to be very different from my manual measurements even though the myelin segmentation was pretty decent.

Also, in the training there seems to be an issue, the axons are segmented properly, they are considered background for some reason? Is there a way to fix this, or was this an issue with my manual masks.

![Figure_1](https://user-images.githubusercontent.com/50454820/87932344-f702b180-cabd-11ea-9859-6d183f0814ed.png)

**Comment by mathieuboudreau** [comment]
Hi Parth - could you please also send us an image of your manually segmented mask so that we can compare?

--
Mathieu Boudreau, PhD
Research Fellow, Montreal Heart Institute
Software Developer for qMRLab and AxonDeepSeg
438-822-8747 | www.github.com/mathieuboudreau

________________________________

**Comment by Paasu** [comment]
![mask](https://user-images.githubusercontent.com/50454820/87939741-0961f880-ca88-11ea-9641-36d564643e49.png)

**Comment by Stoyan-I-A** [comment]
The manual segmentation looks fine in general. However, I notice that some small spots have pixel values that neither correspond to Axon or Myelin. @mathieuboudreau , @mariehbourget , do you think this could be a problem?

**Comment by mathieuboudreau** [comment]
@Paasu did you use a tool to create this mask? Or is it actually a manual segmentation?

**Comment by Paasu** [comment]
I used photoshop.

**Comment by mariehbourget** [comment]
> The manual segmentation looks fine in general. However, I notice that some small spots have pixel values that neither correspond to Axon or Myelin. @mathieuboudreau , @mariehbourget , do you think this could be a problem?

I don't think that would cause a problem like this with only the segmentation of the axons since the values of the masks are binarized during training.

@Paasu can you send us the original image?

**Comment by Paasu** [comment]
![image](https://user-images.githubusercontent.com/50454820/88136818-a7190d00-cbd9-11ea-8fbf-e469d11fb363.png)

**Comment by mathieuboudreau** [comment]
Hi @Paasu,

We had a group meeting today, and think a major source of the issue with your trained model is the quality of your manual masks. It appears there's a few issues with them (e.g. more than three class values, not smooth, etc), and these might stem from your workflow on creating them with your image editing software (photoshop).

To resolve your issue, we think that the best way forward is for us to create a comprehensive step-by-step guide on how to create good manual masks. We're making this a high-priority task, and @mariehbourget will be working on it. We'll let you know when we have that completed, and will share it with you. You can follow the progress in issue #336

**Comment by Paasu** [comment]
Hey @mathieuboudreau 

Thank you for answering my question. I'll keep an eye on the guide to masking. Does this issue with my automatic segmentation also contribute to the disparity in measurements of the myelin thickness between the manual measurement and using the Morphometrics notebook?

**Comment by mariehbourget** [comment]
Hi @Paasu 
The guide to create manual masks is completed. You can find it in our documentation [here](https://axondeepseg.readthedocs.io/en/latest/documentation.html#guide-for-manual-labelling).
To answer your question, the morphometrics in the notebook are extracted from the automatic segmentation. Both axons and myelin must be segmented correctly for the morphometrics to compute properly. Therefore, it is necessary to first fix the issue with the trained model to avoid disparities with your manual measurements.

**Comment by mathieuboudreau** [comment]
Just to hope onto what @mariehbourget , I would recommend that you try creating masks using the GIMP instructions, since the FSLeyes one assumes that your data is compatible with our segmentation models. If you want, you could post an example image here after you've labelled it and we can give you (@Paasu) feedback on the quality.

**Comment by Paasu** [comment]
@mathieuboudreau 
Hello, I have created the new manual masks for a new set of training. 
[Masks.zip](https://github.com/neuropoly/axondeepseg/files/5270051/Masks.zip)
Please let me know if they are optimal for training or need any additional files.

**Comment by mariehbourget** [comment]
Hi @Paasu 
The new masks you created should be good for training. I would maybe try to smooth the borders between structures for mask image2.png like you did for the other ones. 

Also, on some edges in your masks, there are still intermediate values that are not 0, 127 of 255.
Example:
![Capture_edges](https://user-images.githubusercontent.com/54086142/94194111-3b744a00-fe7f-11ea-99a7-a5368c0bbb35.PNG)

However, this is minor and should not impact the training at this point. We suggest you try to do the training as is. Just keep that in mind if you have other masks to create in the future.
Let us know if you have further questions.

**Comment by Paasu** [comment]
Thanks for the prompt reply @mariehbourget 

I have been making changes to some of the images and attempted multiple training sessions but have failed to achieve accurate results. This time I attempted 300 epochs and swapped out image2 with a more accurate image.

![image](https://user-images.githubusercontent.com/50454820/94826148-8ac5f780-03f6-11eb-9cfd-d4d2ac534225.png)
![AxonDeepSeg](https://user-images.githubusercontent.com/50454820/94826155-8c8fbb00-03f6-11eb-9e54-08911ff8b73a.png)
Here are the training files used to train: [training files.zip](https://github.com/neuropoly/axondeepseg/files/5312965/training.files.zip)
The myelin segmentation appears to be accurate and the axon segmentation is more accurate than our previous attempt but still fails to be accurate enough to be useful.

What should be the next steps to approach this? A double check of the masks? Or a change in the training file?

**Comment by Stoyan-I-A** [comment]
The team will likely have a discussion about this tomorrow. 
For now, my guess is that the patch size is too small: you are using a patch size of 256 while the axons are much larger than that in this image. This would also explain why there are vertical and horizontal lines in the axon mask.

That being said, we will write another comment tomorrow after our meeting.

**Comment by Paasu** [comment]
Thanks for helping out!
Here are some additional details that may help you out: 
 - Image size: 4008x2672 (pixels) across all images
 - 0.00166666666 micrometres per pixel
Please let me know if you want any other details.

**Comment by mariehbourget** [comment]
Hi @Paasu , thanks for the training files and additional details.

I suspect that the issue might be with the testing part of your script.
In training.py, line 145:
`prediction = axon_segmentation(path_img, file_img, path_model, config_network, acquired_resolution=0.01, resampled_resolutions=0.01, verbosity_level=3)`
The `acquired_resolution` should correspond to the pixel size in micrometer (in your case: 0.001666666). This argument is not required if you have the `pixel_size_in_micrometer.txt` in the same folder as your testing image.
The `resampled_resolutions` should be the same as the `general_pixel_size` in the dataset builder script (in your case: 0.0086).
Can you try the testing part with these settings and your already trained model?

Also, we discussed your issue in meeting today. If the test I suggest above doesn't work, we think that the best way to move forward would be to try the training on our side. Can you send us your training and testing images and masks?

**Comment by Paasu** [comment]
Thanks for the reply @mariehbourget 

In my training_test folder, I have a the `pixel_size_in_micrometer.txt ` so I believe I don't need to make changes there.
I'll make changes to the `resample_resolutions` in the `training.py` but do I also need to change the patch size to 512 from 256? And if so, would I keep the value at 0.0086?

**Comment by mariehbourget** [comment, marked as answer]
You should only change the line 145 in `training.py`.
If the pixel_size_in_micrometer.txt file is present, remove the `acquired_resolution=0.01` parameter.
Set `resample_resolutions` to 0.0086.
Then, try to re-run the testing part (from line 135) with the model you've already trained.

**Comment by Paasu** [comment]
Hi @mariehbourget,
Thank you for the advice please thank the rest of the team for me!
Changing the values worked really well for me, here are the results under 500 epochs:
![image](https://user-images.githubusercontent.com/50454820/94998264-1af37080-05a0-11eb-937d-187c29595ebb.png)
![AxonDeepSeg](https://user-images.githubusercontent.com/50454820/94998265-1c249d80-05a0-11eb-9f3d-91bf260204b0.png)

The accuracy is much better than before!
Is there any other steps I could take to improve the accuracy?

**Comment by mariehbourget** [comment]
Hi @Paasu , this is great news! 😊

We are happy to see that you have good results with your training. It seems that the accuracy is pretty good except for some parts with decompacted myelin.

If these results are as good on other testing images and you want to extract morphometrics, you can always do minor manual corrections on the segmentation results. You can find a procedure with FSLeyes in our documentation [here](https://axondeepseg.readthedocs.io/en/latest/documentation.html#manual-correction-of-segmentation-masks).

As for training optimization. you can try to train with 512x512 patches instead of 256x256 which can help with axons being mostly fully contained within each patch.
However, considering you already have good results, I'm not sure how this would make such a difference.

**Comment by jcohenadad** [comment]
i think we close this issue now

**Comment by Paasu** [comment]
Thanks for the help, I'll look into using 512x512 patches and manual corrections with FSLeyes @mariehbourget. 
I agree with @jcohenadad, thanks again for the help!
